test: Remove --noshutdown flag, Tidy startup failures

maflcko opened the pull request:
The `--noshutdown` flag is brittle, confusing, and redundant:

* Someone wanting to inspect the state after a test failure will likely also want to debug the state on the python side, so the option is redundant with `--pdbonfailure`. If there was a use case to replicate `--pdbonfailure` without starting pdb, a dedicated flag could be added for that use case.
* It is brittle to use the flag for a passing test, because it will disable checks in the test. For example, on shutdown LSan will perform a leak check, and the test framework will check that the node did not crash, and it will check that the node did not print errors to stderr.

Fix all issues by removing it.

Also, tidy up startup error messages to be less confusing as a result.

Commit:
test: Treat leftover process as error

Printing to stderr instead of stdout makes the test_runner.py fail on
leftover processes. This is desired and fine, because a leftover process
should only happen on a test failure anyway.

Commit:
test: Remove --noshutdown flag

Commit:
test: Avoid redundant stop and error spam on startup failure

Trying to immediately shut down a node after a startup failure without
waiting for the RPC to be fully up will in most cases just fail and lead
to an RPC error.

Also, it is confusing to sidestep the existing fallback to kill any
leftover nodes on a test failure.

So just rely on the fallback.

pablomartin4btc reviewed (commented):
cr ACK and tACK fae3bf6b870eb0f9cddd1adac82ba72890806ae3 (coming from #31462).

Adding `sys.stderr` was necessary for the CI checks (1st commit - fad441fba07877ea78ed6020fde11828307273a6). Redundant error messages were eliminated (e.g. in the proposed PR's [test](https://github.com/bitcoin/bitcoin/pull/31620#issuecomment-2577352102): "`Error: no RPC connection`" is shown only once instead of twice as in `master`).

Just for reference, `--noshutdown` was added in #6032 and `--pdbonfailure` in #11023.

hodlinator reviewed (commented):
Concept ACK fae3bf6b870eb0f9cddd1adac82ba72890806ae3

Thanks for [digging into the history](https://github.com/bitcoin/bitcoin/pull/31620#pullrequestreview-2538211296) @pablomartin4btc. #6032 from year 2015 added `--noshutdown` in order to change the default behavior of `--nocleanup` from leaving lingering *bitcoind*-processes into shutting them down by default. I suspect no one relies on lingering processes so it makes sense to remove it.

On a related note I just opened draft PR #31723 which makes the test framework tell *bitcoind* to pause during startup until one attaches a C++ debugger. That way one can have a debugging session running from the beginning of the test instead of attaching to a lingering process after the test already failed like with `--noshutdown`.

Good to avoid knock-on errors, similarly to #30660 but more minimal. Contrary to that one, this PR doesn't add any tests, but I repurposed one from the other PR and found a way to further tidy up the output (see inline comment).

Commit:
test: Avoid redundant stop and error spam on shutdown

Trying to shut down a node after a test failure may fail and lead to an
RPC error.

Also, it is confusing to sidestep the existing fallback to kill any
leftover nodes on a test failure.

So just rely on the fallback.

Idea by Hodlinator.

Co-Authored-By: Hodlinator <172445034+hodlinator@users.noreply.github.com>

hodlinator reviewed (approved):
ACK fa0369cccf7c2cde244ec6d8669605a5814e4a0a

Thanks for taking my suggestion! (Wouldn't mind being official co-author of that last commit, but it's basically just an extra condition ;) ).

---

Didn't fully realize it before but without that last commit one didn't get this output block because of the exception inside `shutdown()`:
```
TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_bkz3vya4
TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_bkz3vya4/test_framework.log
TestFramework (ERROR): 
TestFramework (ERROR): Hint: Call /home/hodlinator/b2/test/functional/combine_logs.py '/tmp/bitcoin_func_test_bkz3vya4' to consolidate all logs
TestFramework (ERROR): 
TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log.
TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
TestFramework (ERROR): 
```
So it doesn't just prevent spam, but actually fixes this block not being printed.

---

Tested through [recommended wallet_migration.py-method](https://github.com/bitcoin/bitcoin/pull/31620#issuecomment-2577352102).

hodlinator reviewed (approved):
re-ACK faf2f2c654d9aa18b2f49a157956f9ab0fce302a

Thanks for the attribution! Confirmed via range-diff to only be a commit message change.

pablomartin4btc reviewed (commented):
re tACK faf2f2c654d9aa18b2f49a157956f9ab0fce302a

> Didn't fully realize it before but without that last commit one didn't get this output block because of the exception inside shutdown():

> So it doesn't just prevent spam, but actually fixes this block not being printed.

Since the PR hasn't been rebased I tested that running (_edited_: with the `v28.0` folder and its binaries in `releases/`)  `wallet_migration.py` (fixed in #31451).

If the PR gets rebased I'll consider this:

> Just as a test-note: The steps may not work when this pull is rebased on master, so reverting #31462 may be helpful if someone wishes to. It should be possible to come up with other steps to test, but I think having one way to test is enough for now.
